prototype_source/pt2e_quant_qat_static.rst

```diff
@@ -0,0 +1,429 @@
+    num_epochs_between_evals = 2
+    
+    # QAT takes time and one needs to train over a few epochs.
+    # Train and check accuracy after each epoch
+    for nepoch in range(num_epochs):
+        train_one_epoch(prepared_model, criterion, optimizer, data_loader, "cuda", num_train_batches)
+
+        # Optionally disable observer/batchnorm stats after certain number of epochs
+        if epoch >= num_observer_update_epochs:
+            print("Disabling observer for subseq epochs, epoch = ", epoch)
+            prepared_model.apply(torch.ao.quantization.disable_observer)
+        if epoch >= num_batch_norm_update_epochs:
+            print("Freezing BN for subseq epochs, epoch = ", epoch)
+            for n in prepared_model.graph.nodes:
+                # Args: input, weight, bias, running_mean, running_var, training, momentum, eps
+                # We set the `training` flag to False here to freeze BN stats
+                if n.target in [
+                    torch.ops.aten._native_batch_norm_legit.default,
+                    torch.ops.aten.cudnn_batch_norm.default,
+                ]:
+                    new_args = list(n.args)
+                    new_args[5] = False
+                    n.args = new_args
+            prepared_model.recompile()
+    
+        # Check the quantized accuracy every N epochs
+        # Note: If you wish to just evaluate the QAT model (not the quantized model),
+        # then you can just call `torch.ao.quantization.move_exported_model_to_eval/train`.
+        # However, the latter API is not ready yet and will be available in the near future.
+        if (nepoch + 1) % num_epochs_between_evals == 0:
+            prepared_model_copy = copy.deepcopy(prepared_model)
+            quantized_model = convert_pt2e(prepared_model_copy)
+            top1, top5 = evaluate(quantized_model, criterion, data_loader_test, neval_batches=num_eval_batches)
+            print('Epoch %d: Evaluation accuracy on %d images, %2.2f' % (nepoch, num_eval_batches * eval_batch_size, top1.avg))
+
+
+Convert the Trained Model to a Quantized Model
+----------------------------------------------------------
+
+``convert_pt2e`` takes a calibrated model and produces a quantized model.
+Note that, before inference, you must first call
+`torch.ao.quantization.move_exported_model_to_eval()` to ensure certain ops
+like dropout behave correctly in the eval graph. Otherwise, we would continue
+to incorrectly apply dropout in the forward pass during inference, for example.
+
+.. code-block:: python
+
+    quantized_model = convert_pt2e(prepared_model)
+
+    # move certain ops like dropout to eval mode, equivalent to `m.eval()`
+    torch.ao.quantization.move_exported_model_to_eval(m)
+
+    print(quantized_model)
+
+    top1, top5 = evaluate(quantized_model, criterion, data_loader_test, neval_batches=num_eval_batches)
+    print('Final evaluation accuracy on %d images, %2.2f' % (num_eval_batches * eval_batch_size, top1.avg))
```

Comment:
should we show the results here? and compare with baseline?

Reply:
I don't have the results right now. Can we land this first for the 2.1 blog post and add it later?

Reply:
OK then maybe add a comment?